blocks_vertical/procedures.js

```diff
@@ -28,6 +28,289 @@ goog.provide('Blockly.Blocks.procedures');
 goog.require('Blockly.Blocks');
 goog.require('Blockly.constants');
 
+// TODO: Create a namespace properly.
+Blockly.ScratchBlocks.ProcedureUtils = {};
+
+/**
+ * Returns the name of the procedure this block calls, or the empty string if
+ * it has not yet been set.
+ * @return {string} Procedure name.
+ * @this Blockly.Block
+ */
+Blockly.ScratchBlocks.ProcedureUtils.getProcCode = function() {
+  return this.procCode_;
+};
+
+/**
+ * Create XML to represent the (non-editable) name and arguments.
+ * @return {!Element} XML storage element.
+ * @this Blockly.Block
+ */
+Blockly.ScratchBlocks.ProcedureUtils.callerMutationToDom = function() {
+  var container = document.createElement('mutation');
+  container.setAttribute('proccode', this.procCode_);
+  container.setAttribute('argumentnames', JSON.stringify(this.argumentNames_));
+  container.setAttribute('argumentdefaults', JSON.stringify(this.argumentDefaults_));
+  container.setAttribute('argumentids', JSON.stringify(this.argumentIds_));
+  container.setAttribute('warp', this.warp_);
+  return container;
+};
+
+/**
+ * Parse XML to restore the (non-editable) name and parameters.
+ * @param {!Element} xmlElement XML storage element.
+ * @this Blockly.Block
+ */
+Blockly.ScratchBlocks.ProcedureUtils.callerDomToMutation = function(xmlElement) {
+  this.procCode_ = xmlElement.getAttribute('proccode');
+  this.argumentNames_ =  JSON.parse(xmlElement.getAttribute('argumentnames'));
+  this.argumentDefaults_ =  JSON.parse(xmlElement.getAttribute('argumentdefaults'));
+  this.argumentIds_ = JSON.parse(xmlElement.getAttribute('argumentids'));
+  this.warp_ = xmlElement.getAttribute('warp');
+  this.updateDisplay_();
+};
+
+/**
+ * Remove all inputs on the block, including dummy inputs.
+ * @private
+ * @this Blockly.Block
+ */
+Blockly.ScratchBlocks.ProcedureUtils.removeAllInputs_ = function() {
+  // remove all inputs, including dummy inputs.
+  for (var i = 0, input; input = this.inputList[i]; i++) {
+    if (input.connection && input.connection.targetBlock()) {
+      console.warn("connection was still attached?!");
+    }
+    input.dispose();
+  }
+  this.inputList = [];
+};
+
+/**
+ * Disconnect old blocks from all value inputs on this block, but hold onto them
+ * in case they can be reattached later.
+ * @return {!Object.<string, Blockly.Block>} An object mapping parameter IDs to
+ *     the blocks that were connected to those IDs at the beginning of the
+ *     mutation.
+ * @private
+ * @this Blockly.Block
+ */
+Blockly.ScratchBlocks.ProcedureUtils.disconnectOldBlocks_ = function() {
+  // Remove old stuff
+  var connectionMap = {};
+  for (var id in this.paramMap_) {
+    var input = this.paramMap_[id];
+    console.log(id + ' ' + input);
```

Comment:
Do we want to leave the logging in? (ditto for other places)

Reply:
Done.